sources/driverdog/src/main.rs

```diff
@@ -0,0 +1,438 @@
+
+/// Path where the linked kernel modules will be created
+const LIB_MODULES_PATH: &str = "/lib/modules";
+/// Path to the kernel sources
+const KERNEL_SOURCES: &str = "/usr/src/kernels";
+/// Path to the uname bin
+const UNAME_BIN_PATH: &str = "/usr/bin/uname";
+
+/// Paths to kmod utilities
+const DEPMOD_BIN_PATH: &str = "/usr/bin/depmod";
+const MODPROBE_BIN_PATH: &str = "/usr/bin/modprobe";
+
+/// Paths to binutils tools
+const LD_BIN_PATH: &str = "/usr/bin/ld";
+const STRIP_BIN_PATH: &str = "/usr/bin/strip";
+
+/// Stores arguments
+#[derive(FromArgs, PartialEq, Debug)]
+struct Args {
+    /// log-level trace|debug|info|warn|error
+    #[argh(option)]
+    log_level: Option<LevelFilter>,
+    /// configuration file with the modules to be linked
+    #[argh(
+        option,
+        default = "DEFAULT_DRIVER_CONFIG_PATH.to_string()",
+        short = 'd'
+    )]
+    driver_config_path: String,
+    #[argh(subcommand)]
+    subcommand: Subcommand,
+    /// the modules set used to operate
+    #[argh(option)]
+    modules_set: Option<String>,
+}
+
+/// Stores the subcommand to be executed
+#[derive(FromArgs, Debug, PartialEq)]
+#[argh(subcommand)]
+enum Subcommand {
+    LinkModules(LinkModulesArgs),
+    LoadModules(LoadModulesArgs),
+}
+
+/// Links the kernel modules
+#[derive(FromArgs, Debug, PartialEq)]
+#[argh(subcommand, name = "link-modules")]
+struct LinkModulesArgs {}
+
+/// Loads the kernel modules
+#[derive(FromArgs, Debug, PartialEq)]
+#[argh(subcommand, name = "load-modules")]
+struct LoadModulesArgs {}
+
+/// Holds the configuration used to link kernel modules
+#[derive(Deserialize, Debug)]
+struct DriverConfig {
+    #[serde(rename(deserialize = "lib-modules-path"))]
+    lib_modules_path: String,
+    #[serde(rename(deserialize = "objects-source"))]
+    objects_source: String,
+    #[serde(rename(deserialize = "kernel-modules"))]
+    kernel_modules: HashMap<String, Linkable>,
+    #[serde(rename(deserialize = "object-files"))]
+    object_files: HashMap<String, Linkable>,
+}
+
+/// Holds the objects to be linked for the object/kernel module
+#[derive(Deserialize, Debug)]
+struct Linkable {
+    #[serde(rename(deserialize = "link-objects"))]
+    link_objects: Vec<String>,
+}
+
+// Links the modules in the modules sets
+fn link_modules_sets(
+    modules_sets: &HashMap<String, DriverConfig>,
+    target: Option<String>,
+) -> Result<()> {
+    // Get current kernel version
+    let kernel_version = get_kernel_version()?;
+
+    // If the target module set was given, link the kernel modules in it
+    if let Some(target) = target {
+        let driver_config = modules_sets
+            .get(&target)
+            .context(error::MissingModuleSet { target })?;
+        link_modules(&driver_config, &kernel_version)?
+    }
+
+    // Link all the modules sets if no target module was given
+    for driver_config in modules_sets.values() {
+        link_modules(&driver_config, &kernel_version)?;
+    }
```

Comment:
Looks like we always link all module sets, even if we pass the option for a specific set. The load modules function has the same issue.